### PR TITLE
[Login- "Enter Your Store Address" flow] - Add `Sign in with site credentials` option. 

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Related to https://github.com/woocommerce/woocommerce-ios/issues/7298
Design ref: p1658138064625899/1658117084.869099-slack-C0354HSNUJH

### Description
This PR adds ` Site in with site credentials` option to `GetStartedViewController` when trying to sign in using the "Enter your Store Address" flow. 

This is part of "**Experiment 4: Sign in with site credentials (iOS)**" in pe5sF9-6C-p2

### What is changed?

`GetStartedViewController`
- Introduced `ScreenMode` to handle WPCOM and .org site credentials login flows.
- Hidden social buttons in WPCOM mode.
- Moved continue button to the bottom of the screen in WPCOM mode and keep it visible when keyboard appears.
- Add "Sign in with site credentials" button for .org site credentials flow.
- Remove hard coded height values from storyboard and use stackview and autolayout to calculate height. This fixes unnecessary scroll behaviour in small screen even when there is enough space.
- Adjust stackview layout margins to get vertical space and stop using `SpacerTableViewCell`.

`SiteCredentialsViewController`
- Create LoginFields using the response from XML-RPC .org site credentials sign in response.
- If `isWordPressComCredentialsRequired` is true ask user to login using WPCOM credentials by presenting unified password view with `jetpack_user_email` prefilled.

`WordPressOrgCredentials.swift`
- Add a `siteURL` property to strip xmlrpc path from site URL.

`WPStyleGuide+Login.swift`
- Add a helper to generate atrributed string for "Sign in with site credentials" button

### Testing instructions 

- [ ] Check out the branch from https://github.com/woocommerce/woocommerce-ios/pull/7320 and follow the testing instructions specified in the PR. 
- [ ] Check out the branch from https://github.com/wordpress-mobile/WordPress-iOS/pull/19090 and follow the testing instructions specified in the PR. 

### Screenshots

| Before  | After |
| ------------- | ------------- |
| Before |  After |

### CODE REVIEW

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift [hunk 1]
```diff
@@ -95,6 +95,16 @@ public struct WordPressAuthenticatorConfiguration {
     /// Default value is disabled
     let continueWithSiteAddressFirst: Bool
 
+    /// If enabled shows a "Sign in with site credentials" button in `GetStartedViewController` when landing in the screen after entering site address
+    ///  Used in WooCommerce iOS to enable sign in using site credentials feature.
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Nit: I think we may not need to include that this is used for WCiOS, as any app should be able to use it. Let reword this to something more generic:
```suggestion
    ///  Used to enable sign-in to self-hosted sites using WordPress.org credentials.
```

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift [hunk 2]
```diff
@@ -95,6 +95,16 @@ public struct WordPressAuthenticatorConfiguration {
...
+    ///  Disabled by default
+    let enableSiteCredentialsLoginInGetStartedScreen: Bool
+
+    /// If enabled, we will ask for WPCOM login after signing in using .org site credentials.
+    ///  WordPress iOS app can proceed without WPCOM login. But, WooCommerce iOS app needs WPCOM login.
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Nit: This line is also too specific for a library, let's remove this.

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift [hunk 3]
```diff
@@ -95,6 +95,16 @@ public struct WordPressAuthenticatorConfiguration {
     /// Default value is disabled
     let continueWithSiteAddressFirst: Bool
 
+    /// If enabled shows a "Sign in with site credentials" button in `GetStartedViewController` when landing in the screen after entering site address
+    ///  Used in WooCommerce iOS to enable sign in using site credentials feature.
+    ///  Disabled by default
+    let enableSiteCredentialsLoginInGetStartedScreen: Bool
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Nit: We might not need to mention where the button is, but instead what the button is for:
```suggestion
    let enableSiteCredentialsLoginForSelfHostedSites: Bool
```

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift [hunk 4]
```diff
@@ -26,18 +41,26 @@ class GetStartedViewController: LoginViewController {
...
-        button.addTarget(self, action: #selector(handleSubmitButtonTapped(_:)), for: .touchUpInside)
-        button.accessibilityIdentifier = "Get Started Email Continue Button"
-
-        let title = WordPressAuthenticator.shared.displayStrings.continueButtonTitle
-        button.setTitle(title, for: .normal)
-        button.setTitle(title, for: .highlighted)
+        button.addTarget(self, action: #selector(handleSubmitButtonTapped), for: .touchUpInside)
+        button.accessibilityIdentifier = ButtonConfiguration.Continue.accessibilityIdentifier
+        button.setTitle(ButtonConfiguration.Continue.title, for: .normal)
+        button.setTitle(ButtonConfiguration.Continue.title, for: .highlighted)
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift]: Nit: Do we really need to set the title for the highlighted mode? I thought if we don't set a value to it, the button will use the same one for the `.normal` state.

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift [hunk 5]
```diff
@@ -734,3 +830,30 @@ extension GetStartedViewController: UITextFieldDelegate {
...
+    }
+}
+
+// MARK: - Button configuration
+
+private extension GetStartedViewController {
+    enum ButtonConfiguration {
+        enum Continue {
+            static let title = WordPressAuthenticator.shared.displayStrings.continueButtonTitle
+            static let accessibilityIdentifier = "Get Started Email Continue Button"
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift:862]: Nit: For accessibility identifiers, I think we want to follow the naming style for IDs instead:
```suggestion
            static let accessibilityIdentifier = "get-started-email-continue-button"
```

But since the naming is not consistent in this library, this is just a suggestion.

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift [hunk 6]
```diff
@@ -734,3 +830,30 @@ extension GetStartedViewController: UITextFieldDelegate {
...
+        }
+
+        enum SignInWithSiteCredentials {
+            static let accessibilityIdentifier = "Sign in with site credentials Button"
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift:866]: Similar suggestion as above:
```suggestion
            static let accessibilityIdentifier = "sign-in-with-site-credentials-button"
```

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift [hunk 7]
```diff
@@ -95,6 +95,16 @@ public struct WordPressAuthenticatorConfiguration {
...
+
+    /// If enabled, we will ask for WPCOM login after signing in using .org site credentials.
+    ///  WordPress iOS app can proceed without WPCOM login. But, WooCommerce iOS app needs WPCOM login.
+    ///  Disabled by default
+    let isWordPressComCredentialsRequired: Bool
```
Comment (itsmeichigo) [inline WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Nit: I'm thinking we can also clarify this further:
```suggestion
    let isWPComLoginRequiredForSiteCredentialsLogin: Bool
```

**Review by itsmeichigo — COMMENTED** [review]
Great work @selanthiraiyan 👏 This works nicely with https://github.com/woocommerce/woocommerce-ios/pull/7320 - but one thing stood out to me: the Get Started screen looks a bit off compared to the design:

| Design | Reality |
| ----- | ----- |
|  |  |

As you can see above, the icon in the "Sign in with store credentials" button has a slightly bolder color compared to the design. And the "What is WordPress.com" button has an underline, which looks more like a hyperlink on a web page than a button on a native app. 😅

When I was testing, I navigated back to change my site address, but I was stuck on the password screen as the back button disappears as soon as I tapped the "Continue" button:



I also left a few nit-picking comments below.

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/98e80ab0828e6c3398fb7735e56200f7dc5e714f

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/9dd2d8550566e918534d16f54b0579232fcb417f

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/5e7885880c8d902a807511de2b418fdb8d4972e4

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift]: Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/1d2641bdb8b0ffaf5972dde7eb2509c1e4dafe74

File: WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift (hunk 7 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Authenticator/WordPressAuthenticatorConfiguration.swift]: Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/a5b3874a6efcdc2eae4cdcccc4e7e97fe51e9718

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift (hunk 5 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift:862]: I would like to keep using the old name as I only refactored and moved this code here. And, I don't want to break any tests that might be using this.

File: WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift (hunk 6 above)
Comment (selanthiraiyan) [reply WordPressAuthenticator/Unified Auth/View Related/Get Started/GetStartedViewController.swift:866]: To keep it consistent with [the other style](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649#discussion_r929609931) I would like to keep the existing style.

**Comment by selanthiraiyan** [comment]
> As you can see above, the icon in the "Sign in with store credentials" button has a slightly bolder color compared to the design.

Fixed in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/85ff5ca83983e6440862952fde320e4da279e638

**Comment by selanthiraiyan** [comment]
> And the "What is WordPress.com" button has an underline, which looks more like a hyperlink on a web page than a button on a native app.

Checked with the design team and updated in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/e3da8d122ade23498774732be21558b59d71ac91

Design confirmation here - p1658816607458229/1658117084.869099-slack-C0354HSNUJH . Thanks @annchichi 

Attaching the latest screenshot below,
UpdatedScreen

**Comment by selanthiraiyan** [comment]
> When I was testing, I navigated back to change my site address, but I was stuck on the password screen as the back button disappears as soon as I tapped the "Continue" button:

Great catch. 🎉 

Fixed in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/b8c95a0c68c822a11e177ee62680b47842ec19f3

**Comment by selanthiraiyan** [comment]
Thank you for the detailed review, @itsmeichigo. 🙇 

I have included the necessary changes, and the PR is ready for another look now.

**Review by itsmeichigo — APPROVED** [review]
Thank you @selanthiraiyan for the update! The feature works nicely now!

I read the discussion in p1658816607458229/1658117084.869099-slack-C0354HSNUJH, and I see that Ann mentioned the "What is WordPress.com?" button to be the same size as "Find your store address" button. 

From what I see the size aren't really identical at the moment, with the "What is WordPress.com" button a little bigger. I took a look at the storyboard for `LoginSiteAddressViewController` and it seems that the "Find your store address" button uses `subHead` font. Maybe we should use the same font for consistency?

Other than that, I think this PR is all good! 💯

**Comment by selanthiraiyan** [comment]
> From what I see the size aren't really identical at the moment, with the "What is WordPress.com" button a little bigger. I took a look at the storyboard for LoginSiteAddressViewController and it seems that the "Find your store address" button uses subHead font. Maybe we should use the same font for consistency?

Done in https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/649/commits/87445743d1ec3c69f0c56a320681474b1fb9bf53
